### DISCUSSION TITLE
esp32 and miel hvacs: MiElHVAC: checksum failed, restarting

### DISCUSSION THREAD

**Discussion by host-felice** [arendst/Tasmota#19239] [Support request]
Hello, I am using an esp32 with the latest development 13.0.0.3 connected to miel hvacs (mitsubishi air conditioner). The esp32 module is a wemos d1 mini esp32 clone. I used the template for this module. Seriallog 0. Between the esp32 and the AC there is a level shifter.

I first used gpio1 and gpio3 but on the d1 mini esp32 it is not recommended so now I am using gpio16 (rx) and gpio17 (tx). Both with tx/rx and with gpio16/gpio17 with weblog 4 I get some MiElHVAC: checksum failed, restarting. The air conditioner seems to work fine, but with the previous esp8266 module I never had this error:

16:56:50.469 MiElHVAC: sending FC42013010 02000000000000000000000000000000 7b
16:56:50.677 MiElHVAC: sending FC42013010 03000000000000000000000000000000 7a
16:56:50.831 MiElHVAC: checksum failed, restarting
16:56:50.934 MiElHVAC: sending FC42013010 02000000000000000000000000000000 7b
16:56:51.208 MiElHVAC: response data 0200000001050000000083B446000000
16:56:51.217 MiElHVAC: sending FC42013010 09000000000000000000000000000000 74
16:56:51.425 MiElHVAC: sending FC42013010 02000000000000000000000000000000 7b
16:56:51.578 MiElHVAC: checksum failed, restarting
16:56:51.683 MiElHVAC: sending FC42013010 06000000000000000000000000000000 77
16:56:51.936 MiElHVAC: response data 060000000001E70FDC00004200000000
16:56:51.941 MiElHVAC: sending FC42013010 02000000000000000000000000000000 7b
16:56:52.203 MiElHVAC: response data 0200000001050000000083B446000000
16:56:52.213 MiElHVAC: sending FC42013010 03000000000000000000000000000000 7a
16:56:52.466 MiElHVAC: response data 0300000F00C6B3B7FE42000095EC0000

Should I be worried or can I just ignore it?

Thank you

**Comment by barbudor** [comment]
It could be the quality of the ESP32 mini-D1 generating errors
If you have another one, preferably an original one, or an esp32c3 one, that could be worth testing

**Comment by host-felice** [comment]
Unfortunately no other esp32. I bought them so I can use the PCB I made for the D1 mini esp8266 since they have compatible pinout. If there is nothing I can test/change software side, I will need to buy new esp32 and check

**Reply by barbudor** [reply]
I have 1 setup based on a mini-D1 esp32 receiving serial data, not related to AC at all, using a different driver, and I also have a lot of serial errors
I always thought that it was the circuit that feed the serial but that makes me wonder if the module are just not crappy....

**Reply by host-felice** [reply]
well, it could be possible, but isn't the esp32 chip the same? I would expect some problem with the LDO or the antenna... plus I am using a level shifter and not some resistors to communicate with the ac that works @5V

**Reply by sfromis** [reply]
From what I understand, you've wired the HVAC to the serial pins of the ESP32 via a level shifter? In that case, have you verified that the unit is not using something like RS232 (which is common for external serial interfaces)? That would not be electrically compatible, with or without level shifters. ESP32 and ESP8266 may not deal with bad electrics the same way.

**Reply by host-felice** [reply]
thanks @sfromis mitsubishi ac are fully compatible with esp. The AC works at 5V, the esp works at 3,3V. The level shifter is to make the communication possible. You could use a bunch of resistors like this configuration
![](https://raw.githubusercontent.com/SwiCago/HeatPump/master/CN105_ESP8266.png)

but I prefer to use a logic level shifter for a cleaner job. With the esp8266 (wemos d1 mini) I never saw a checksum failed. It happens only with the ESP32.

**Reply by pkkrusty** [reply]
I ran into an issue where I needed to use a pull-up resistor to receive MISO data correctly on a SPI input with ESP32, and with the same hardware setup, never needed to with ESP8266. May be a similar thing, where it's more sensitive to noise or common ground or something. What if you drop the resistors for Rx/Rx to 100k? Still errors?

**Reply by host-felice** [reply]
I didn't try the pull-up resistors at the moment. I will try later today and report. Honestly it is not a big deal some checksum errors cause the esp pulls data way too fast from the AC (and I don't understand why). According with the code, it should be every 30 seconds, but looking at the web console it is more like 200 ms or so.

**Comment by host-felice** [comment]
@barbudor I have a Pzem016 laying around. I will try to connect an esp8266 and one of these esp32 and see if I get some checksum errors.

**Reply by host-felice** [reply]
I tried with the same ESP32 and a PZEM016 and I saw nothing wrong. It worked perfectly and the weblog 4 showed no errors. The PZEM016 is not wired with a level shifter, but with two 10k Ohm resistors like the image I posted for the Mitsubishi AC.

I will try tomorrow to arrange a test with two resistors with the Mitsubishi instead of the level shifter. Just to be sure of the level shifter...

**Comment by host-felice** [comment]
@pkkrusty I have tried with both 10k and 100k resistors, no level shifter, but I still get checksum failed using the esp32.

What I really don't understand is the poll interval. It seems very fast according with the weblog. Maybe slowing down the polls will fix the checksum failed?

**Reply by sfromis** [reply]
To me, 4 times a second  does not look "very fast", but if you want to experiment, you could move the processing from FUNC_EVERY_250_MSECOND to FUNC_EVERY_SECOND.
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_44_miel_hvac.ino#L1315-L1326

**Reply by host-felice** [reply]
I didn't say it is very fast, but that it seems to be very fast. I am not into coding, nor do I understand the hardware limits. I am just trying to understand what the problem is. Since there are 4 polls every second, it is not a big deal to lose some polls from time to time.
